adapter/memQuota/rollingWindow.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memQuota
+
+import (
+	"sync"
+)
+
+// A single allocation, carrying its amount and expiration time
+type alloc struct {
+	next           *alloc
+	count          int64
+	expirationTick int32
+}
+
+var allocPool = sync.Pool { New: func() interface{} { return &alloc{}}}
+
+// Implements a rolling window that allows N units to be allocated per rolling time interval.
+// Time is abstracted in terms of ticks, provided by the caller, decoupling the
+// implementation from real-time, enabling much easier testing and more flexibility.
+type rollingWindow struct {
+	// first node in the list of allocs, or nil when no allocs
+	first *alloc
+
+	// last node in the list of allocs, or undefined when first is nil
+	last *alloc
+
+	// total amount left in the window
+	avail int64
+
+	// determines how many ticks individual allocs survive before being considered dead
+	ticksInWindow int32
+}
+
+func newRollingWindow(limit int64, ticksInWindow int32) *rollingWindow {
+	return &rollingWindow{
+		avail:         limit,
+		ticksInWindow: ticksInWindow,
+	}
+}
+
+func (w *rollingWindow) alloc(amount int64, currentTick int32) bool {
+	// reap old entries
+	for n := w.first; n != nil && n.expirationTick <= currentTick; n = n.next {
```

Comment:
any reason not to use "container/list" (or similar) for the window/allocs here (as opposed to having the allocs maintain the pointers themselves)?